index.js

```diff
@@ -13,140 +13,144 @@ const defaultOptions = {
   allowedHeaders: null,
   maxAge: null,
   preflight: true,
-  hideOptionsRoute: true,
   strictPreflight: true
 }
 
 function fastifyCors (fastify, opts, next) {
-  const {
-    origin,
-    credentials,
-    exposedHeaders,
-    allowedHeaders,
-    methods,
-    maxAge,
-    optionsSuccessStatus,
-    preflight,
-    hideOptionsRoute,
-    strictPreflight,
-    preflightContinue
-  } = Object.assign({}, defaultOptions, opts)
-
-  const resolveOriginOption = typeof origin === 'function' ? resolveOriginWrapper : (_, cb) => cb(null, origin)
-
   fastify.decorateRequest('corsPreflightEnabled', false)
-  fastify.addHook('onRequest', onRequest)
-
-  if (preflight === true) {
-    // The preflight reply must occur in the hook. This allows fastify-cors to reply to
-    // preflight requests BEFORE possible authentication plugins. If the preflight reply
-    // occurred in this handler, other plugins may deny the request since the browser will
-    // remove most headers (such as the Authentication header).
-    //
-    // This route simply enables fastify to accept preflight requests.
-    fastify.options('*', { schema: { hide: hideOptionsRoute } }, (req, reply) => {
-      if (!req.corsPreflightEnabled) {
-        // Do not handle preflight requests if the origin option disabled CORS
-        reply.callNotFound()
-        return
-      }
 
-      reply.send()
-    })
+  let optionsCallback
+  let hideOptionsRoute = true
+  if (typeof opts === 'function') {
+    optionsCallback = opts
+  } else {
+    hideOptionsRoute = opts.hideOptionsRoute
+    optionsCallback = function (req, cb) {
+      cb(null, opts)
+    }
   }
+  fastify.addHook('onRequest', (req, reply, next) => {
+    optionsCallback(req, (err, options) => {
```

Comment:
We should also support promises.

Reply:
@mcollina Do you mean that there should be another option to provide calculated cors options?
Like
```
const options = await optionsCallback(request)
```

Reply:
I honestly don't know how to achieve this, I would have to somehow check if the function expects callback as the second argument or if it returns something. I don't know if it can be handled somehow

Reply:
You can take this as an example: https://github.com/fastify/fastify/blob/master/lib/hooks.js#L142-L148